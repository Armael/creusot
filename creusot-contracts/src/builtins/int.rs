use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int;

impl PartialEq for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "eq_int"]
    fn eq(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ne_int"]
    fn ne(&self, _: &Int) -> bool {
        panic!()
    }
}

impl PartialOrd for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "le_int"]
    fn le(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ge_int"]
    fn ge(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "lt_int"]
    fn lt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "gt_int"]
    fn gt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    fn partial_cmp(&self, _: &Self) -> Option<std::cmp::Ordering> {
        panic!()
    }
}

impl From<i32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "i32_to_int"]
    fn from(_: i32) -> Self {
        panic!()
    }
}

impl From<u32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "u32_to_int"]
    fn from(_: u32) -> Self {
        panic!()
    }
}

impl From<usize> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "usize_to_int"]
    fn from(_: usize) -> Self {
        panic!()
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
