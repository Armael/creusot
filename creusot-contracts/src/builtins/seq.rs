use creusot_contracts_proc::*;

use crate::Int;

#[creusot::builtins = "seq.Seq.seq"]
pub struct Seq<T>(std::marker::PhantomData<T>);

impl<T> Seq<T> {
    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.get"]
    pub fn index(self, _: Int) -> T {
        std::process::abort()
    }

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.length"]
    pub fn len(self) -> Int {
        std::process::abort()
    }

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.set"]
    pub fn set(self, _: Int, _: T) -> Self {
        std::process::abort()
    }

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.snoc"]
    pub fn push(self, _: T) -> Self {
        std::process::abort()
    }
    #[predicate]
    pub fn permutation_of(self, o: Self) -> bool {
        self.permut(o, 0, self.len())
    }

    #[trusted]
    #[predicate]
    #[creusot::builtins = "seq.Permut.permut"]
    pub fn permut(self, _: Self, _: Int, _: Int) -> bool {
        std::process::abort()
    }
}
