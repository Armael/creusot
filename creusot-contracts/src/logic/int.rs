use super::model::*;
use crate as creusot_contracts;
use crate::logic::*;
use creusot_contracts_proc::*;

use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int(*mut ());

impl WellFounded for Int {}

impl Int {
    #[trusted]
    #[logic]
    #[creusot::builtins = "int.Power.power"]
    pub fn pow(self, _: Int) -> Int {
        absurd
    }
}

impl From<u8> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "prelude.UInt8.to_int"]
    fn from(_: u8) -> Self {
        absurd
    }
}
impl Model for u8 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "prelude.UInt8.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<u32> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn from(_: u32) -> Self {
        absurd
    }
}
impl Model for u32 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<u64> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn from(_: u64) -> Self {
        absurd
    }
}
impl Model for u64 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i64> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn from(_: i64) -> Self {
        absurd
    }
}
impl Model for i64 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i32> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn from(_: i32) -> Self {
        absurd
    }
}
impl Model for i32 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<usize> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn from(_: usize) -> Self {
        absurd
    }
}
impl Model for usize {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<isize> for Int {
    #[logic]
    #[trusted]
    #[rustc_diagnostic_item = "isize_to_int"]
    fn from(_: isize) -> Self {
        absurd
    }
}
impl Model for isize {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
