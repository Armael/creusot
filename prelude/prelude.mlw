module CreusotPrelude
  type borrowed 'a = { current : 'a; final: 'a }

  let function ( *_ ) x = x.current
  let function ( ^_ ) x = x.final

  val borrow_mut (a : 'a) : borrowed 'a
    ensures { * result = a }
end
