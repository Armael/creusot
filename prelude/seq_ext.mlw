module SeqExt
use seq.Seq

function subsequence (s : seq 't) (i : int) (j : int) : seq 't = s[i..j]

function mk_empty () : seq 't = Seq.empty
end