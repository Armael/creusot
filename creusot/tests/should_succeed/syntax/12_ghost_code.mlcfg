
module C12GhostCode_GhostArg_Interface
  use mach.int.Int
  use mach.int.UInt32
  val ghost_arg [@cfg:stackify] (ghost g : uint32) : ()
end
module C12GhostCode_GhostArg
  use mach.int.Int
  use mach.int.UInt32
  let rec cfg ghost_arg [@cfg:stackify] [#"../12_ghost_code.rs" 4 0 4 31] (ghost g : uint32) : ()
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  ghost var g_1 : uint32;
  ghost var _x_2 : uint32;
  var _5 : ();
  {
    g_1 <- g;
    goto BB0
  }
  BB0 {
    _5 <- ();
    _x_2 <- ghost ([#"../12_ghost_code.rs" 5 25 5 38] g_1);
    goto BB1
  }
  BB1 {
    _0 <- ();
    return _0
  }
  
end
module Core_Ptr_NonNull_NonNull_Type
  use prelude.Opaque
  type core_ptr_nonnull_nonnull_type 't =
    | Core_Ptr_NonNull_NonNull_Type opaque_ptr
    
end
module Core_Marker_PhantomData_Type
  type core_marker_phantomdata_type 't =
    | Core_Marker_PhantomData_Type
    
end
module Core_Ptr_Unique_Unique_Type
  use Core_Marker_PhantomData_Type
  use Core_Ptr_NonNull_NonNull_Type
  type core_ptr_unique_unique_type 't =
    | Core_Ptr_Unique_Unique_Type (Core_Ptr_NonNull_NonNull_Type.core_ptr_nonnull_nonnull_type 't) (Core_Marker_PhantomData_Type.core_marker_phantomdata_type 't)
    
end
module Alloc_RawVec_RawVec_Type
  use mach.int.Int
  use prelude.UIntSize
  use Core_Ptr_Unique_Unique_Type
  type alloc_rawvec_rawvec_type 't 'a =
    | Alloc_RawVec_RawVec_Type (Core_Ptr_Unique_Unique_Type.core_ptr_unique_unique_type 't) usize 'a
    
end
module Alloc_Vec_Vec_Type
  use mach.int.Int
  use prelude.UIntSize
  use Alloc_RawVec_RawVec_Type
  type alloc_vec_vec_type 't 'a =
    | Alloc_Vec_Vec_Type (Alloc_RawVec_RawVec_Type.alloc_rawvec_rawvec_type 't 'a) usize
    
end
module CreusotContracts_Std1_Vec_Impl0_Model_Interface
  type t
  type a
  use seq.Seq
  use mach.int.Int
  use prelude.UIntSize
  use Alloc_Vec_Vec_Type
  function model (self : Alloc_Vec_Vec_Type.alloc_vec_vec_type t a) : Seq.seq t
end
module CreusotContracts_Std1_Vec_Impl0_Model
  type t
  type a
  use seq.Seq
  use mach.int.Int
  use prelude.UIntSize
  use Alloc_Vec_Vec_Type
  function model (self : Alloc_Vec_Vec_Type.alloc_vec_vec_type t a) : Seq.seq t
  axiom model_spec : forall self : Alloc_Vec_Vec_Type.alloc_vec_vec_type t a . Seq.length (model self) <= 18446744073709551615
end
module Alloc_Alloc_Global_Type
  type alloc_alloc_global_type  =
    | Alloc_Alloc_Global_Type
    
end
module Alloc_Vec_Impl0_New_Interface
  type t
  use seq.Seq
  use mach.int.Int
  use mach.int.Int32
  use Alloc_Alloc_Global_Type
  use Alloc_Vec_Vec_Type
  clone CreusotContracts_Std1_Vec_Impl0_Model_Interface as Model0 with type t = t,
  type a = Alloc_Alloc_Global_Type.alloc_alloc_global_type, axiom .
  val new [@cfg:stackify] (_ : ()) : Alloc_Vec_Vec_Type.alloc_vec_vec_type t (Alloc_Alloc_Global_Type.alloc_alloc_global_type)
    ensures { Seq.length (Model0.model result) = 0 }
    
end
module Alloc_Vec_Impl0_New
  type t
  use seq.Seq
  use mach.int.Int
  use mach.int.Int32
  use Alloc_Alloc_Global_Type
  use Alloc_Vec_Vec_Type
  clone CreusotContracts_Std1_Vec_Impl0_Model_Interface as Model0 with type t = t,
  type a = Alloc_Alloc_Global_Type.alloc_alloc_global_type, axiom .
  val new [@cfg:stackify] (_ : ()) : Alloc_Vec_Vec_Type.alloc_vec_vec_type t (Alloc_Alloc_Global_Type.alloc_alloc_global_type)
    ensures { Seq.length (Model0.model result) = 0 }
    
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve_Interface
  type self
  predicate resolve (self : self)
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve
  type self
  predicate resolve (self : self)
end
module CreusotContracts_Std1_Vec_Impl3_Resolve_Interface
  type t
  use Alloc_Alloc_Global_Type
  use Alloc_Vec_Vec_Type
  predicate resolve (self : Alloc_Vec_Vec_Type.alloc_vec_vec_type t (Alloc_Alloc_Global_Type.alloc_alloc_global_type))
end
module CreusotContracts_Std1_Vec_Impl3_Resolve
  type t
  use Alloc_Alloc_Global_Type
  use Alloc_Vec_Vec_Type
  use mach.int.Int
  use mach.int.Int32
  use seq.Seq
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve_Interface as Resolve0 with type self = t
  clone CreusotContracts_Std1_Vec_Impl0_Model_Interface as Model0 with type t = t,
  type a = Alloc_Alloc_Global_Type.alloc_alloc_global_type, axiom .
  predicate resolve (self : Alloc_Vec_Vec_Type.alloc_vec_vec_type t (Alloc_Alloc_Global_Type.alloc_alloc_global_type)) =
    forall i : int . 0 <= i /\ i < Seq.length (Model0.model self) -> Resolve0.resolve (Seq.get (Model0.model self) i)
end
module CreusotContracts_Std1_Vec_Impl3
  type t
end
module CreusotContracts_Std1_Vec_Impl0
  type t
  type a
end
module CreusotContracts_Logic_Resolve_Impl2_Resolve_Interface
  type t
  predicate resolve (self : t)
end
module CreusotContracts_Logic_Resolve_Impl2_Resolve
  type t
  predicate resolve (self : t) =
    true
end
module CreusotContracts_Logic_Resolve_Impl2
  type t
end
module C12GhostCode_GhostVec_Interface
  val ghost_vec [@cfg:stackify] (_ : ()) : ()
end
module C12GhostCode_GhostVec
  use mach.int.Int
  use mach.int.UInt32
  use Alloc_Alloc_Global_Type
  use Alloc_Vec_Vec_Type
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve1 with type t = uint32
  clone CreusotContracts_Std1_Vec_Impl0_Model as Model0 with type t = uint32,
  type a = Alloc_Alloc_Global_Type.alloc_alloc_global_type, axiom .
  clone CreusotContracts_Std1_Vec_Impl3_Resolve as Resolve0 with type t = uint32, function Model0.model = Model0.model,
  predicate Resolve0.resolve = Resolve1.resolve
  clone Alloc_Vec_Impl0_New_Interface as New0 with type t = uint32, function Model0.model = Model0.model
  let rec cfg ghost_vec [@cfg:stackify] [#"../12_ghost_code.rs" 8 0 8 18] (_ : ()) : ()
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var x_1 : Alloc_Vec_Vec_Type.alloc_vec_vec_type uint32 (Alloc_Alloc_Global_Type.alloc_alloc_global_type);
  ghost var _s_2 : Alloc_Vec_Vec_Type.alloc_vec_vec_type uint32 (Alloc_Alloc_Global_Type.alloc_alloc_global_type);
  var _5 : ();
  {
    goto BB0
  }
  BB0 {
    x_1 <- ([#"../12_ghost_code.rs" 9 22 9 32] New0.new ());
    goto BB1
  }
  BB1 {
    _5 <- ();
    _s_2 <- ghost ([#"../12_ghost_code.rs" 10 32 10 44] x_1);
    goto BB2
  }
  BB2 {
    _0 <- ();
    goto BB3
  }
  BB3 {
    assume { Resolve0.resolve x_1 };
    return _0
  }
  
end
module C12GhostCode_Omg_Interface
  function omg (_ : ()) : ()
end
module C12GhostCode_Omg
  function omg [#"../12_ghost_code.rs" 14 0 14 12] (_ : ()) : () =
    [#"../12_ghost_code.rs" 13 0 13 8] ()
end
module C12GhostCode_GhostCopy_Interface
  val ghost_copy [@cfg:stackify] (_ : ()) : ()
end
module C12GhostCode_GhostCopy
  use mach.int.Int
  use mach.int.Int32
  use seq.Seq
  let rec cfg ghost_copy [@cfg:stackify] [#"../12_ghost_code.rs" 16 0 16 19] (_ : ()) : ()
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var a_1 : int32;
  ghost var _s_2 : Seq.seq int32;
  var _5 : ();
  ghost var _6 : Seq.seq int32;
  var _9 : ();
  {
    goto BB0
  }
  BB0 {
    a_1 <- (0 : int32);
    _5 <- ();
    _s_2 <- ghost ([#"../12_ghost_code.rs" 18 17 18 46] Seq.snoc (Seq.empty ) (0 : int32));
    goto BB1
  }
  BB1 {
    _9 <- ();
    _6 <- ghost ([#"../12_ghost_code.rs" 19 9 19 34] Seq.snoc _s_2 a_1);
    goto BB2
  }
  BB2 {
    _s_2 <- _6;
    _6 <- any Seq.seq int32;
    _0 <- ();
    return _0
  }
  
end
module C12GhostCode_LogiDrop_Interface
  type t
  function logi_drop (_1' : t) : ()
end
module C12GhostCode_LogiDrop
  type t
  function logi_drop [#"../12_ghost_code.rs" 23 0 23 21] (_1' : t) : () =
    [#"../12_ghost_code.rs" 22 0 22 8] ()
end
module Alloc_Vec_Impl1_Push_Interface
  type t
  type a
  use prelude.Borrow
  use seq.Seq
  use Alloc_Vec_Vec_Type
  clone CreusotContracts_Std1_Vec_Impl0_Model_Interface as Model0 with type t = t, type a = a, axiom .
  val push [@cfg:stackify] (self : borrowed (Alloc_Vec_Vec_Type.alloc_vec_vec_type t a)) (value : t) : ()
    ensures { Model0.model ( ^ self) = Seq.snoc (Model0.model ( * self)) value }
    
end
module Alloc_Vec_Impl1_Push
  type t
  type a
  use prelude.Borrow
  use seq.Seq
  use Alloc_Vec_Vec_Type
  clone CreusotContracts_Std1_Vec_Impl0_Model_Interface as Model0 with type t = t, type a = a, axiom .
  val push [@cfg:stackify] (self : borrowed (Alloc_Vec_Vec_Type.alloc_vec_vec_type t a)) (value : t) : ()
    ensures { Model0.model ( ^ self) = Seq.snoc (Model0.model ( * self)) value }
    
end
module Alloc_Vec_Impl1_Len_Interface
  type t
  type a
  use mach.int.UInt64
  use seq.Seq
  use prelude.Borrow
  use mach.int.Int
  use prelude.UIntSize
  use Alloc_Vec_Vec_Type
  clone CreusotContracts_Std1_Vec_Impl0_Model_Interface as Model0 with type t = t, type a = a, axiom .
  val len [@cfg:stackify] (self : Alloc_Vec_Vec_Type.alloc_vec_vec_type t a) : usize
    ensures { UInt64.to_int result = Seq.length (Model0.model self) }
    
end
module Alloc_Vec_Impl1_Len
  type t
  type a
  use mach.int.UInt64
  use seq.Seq
  use prelude.Borrow
  use mach.int.Int
  use prelude.UIntSize
  use Alloc_Vec_Vec_Type
  clone CreusotContracts_Std1_Vec_Impl0_Model_Interface as Model0 with type t = t, type a = a, axiom .
  val len [@cfg:stackify] (self : Alloc_Vec_Vec_Type.alloc_vec_vec_type t a) : usize
    ensures { UInt64.to_int result = Seq.length (Model0.model self) }
    
end
module C12GhostCode_GhostCheck_Interface
  val ghost_check [@cfg:stackify] (_ : ()) : ()
end
module C12GhostCode_GhostCheck
  use mach.int.Int
  use mach.int.Int32
  use prelude.UIntSize
  use prelude.Borrow
  use Alloc_Alloc_Global_Type
  use Alloc_Vec_Vec_Type
  clone CreusotContracts_Logic_Resolve_Impl2_Resolve as Resolve1 with type t = int32
  clone CreusotContracts_Std1_Vec_Impl0_Model as Model0 with type t = int32,
  type a = Alloc_Alloc_Global_Type.alloc_alloc_global_type, axiom .
  clone CreusotContracts_Std1_Vec_Impl3_Resolve as Resolve0 with type t = int32, function Model0.model = Model0.model,
  predicate Resolve0.resolve = Resolve1.resolve
  clone Alloc_Vec_Impl1_Len_Interface as Len0 with type t = int32,
  type a = Alloc_Alloc_Global_Type.alloc_alloc_global_type, function Model0.model = Model0.model
  clone Alloc_Vec_Impl1_Push_Interface as Push0 with type t = int32,
  type a = Alloc_Alloc_Global_Type.alloc_alloc_global_type, function Model0.model = Model0.model
  clone C12GhostCode_LogiDrop as LogiDrop0 with type t = Alloc_Vec_Vec_Type.alloc_vec_vec_type int32 (Alloc_Alloc_Global_Type.alloc_alloc_global_type)
  clone Alloc_Vec_Impl0_New_Interface as New0 with type t = int32, function Model0.model = Model0.model
  let rec cfg ghost_check [@cfg:stackify] [#"../12_ghost_code.rs" 25 0 25 20] (_ : ()) : ()
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var x_1 : Alloc_Vec_Vec_Type.alloc_vec_vec_type int32 (Alloc_Alloc_Global_Type.alloc_alloc_global_type);
  ghost var _2 : ();
  var _4 : ();
  var _5 : ();
  var _6 : borrowed (Alloc_Vec_Vec_Type.alloc_vec_vec_type int32 (Alloc_Alloc_Global_Type.alloc_alloc_global_type));
  var _7 : ();
  var _8 : bool;
  var _9 : bool;
  var _10 : usize;
  var _11 : Alloc_Vec_Vec_Type.alloc_vec_vec_type int32 (Alloc_Alloc_Global_Type.alloc_alloc_global_type);
  var _12 : ();
  {
    goto BB0
  }
  BB0 {
    x_1 <- ([#"../12_ghost_code.rs" 26 16 26 26] New0.new ());
    goto BB1
  }
  BB1 {
    _4 <- ();
    _2 <- ghost ([#"../12_ghost_code.rs" 29 4 29 31] let _ = LogiDrop0.logi_drop x_1 in ());
    goto BB2
  }
  BB2 {
    _6 <- borrow_mut x_1;
    x_1 <-  ^ _6;
    _5 <- ([#"../12_ghost_code.rs" 31 4 31 13] Push0.push _6 (0 : int32));
    goto BB3
  }
  BB3 {
    _11 <- x_1;
    _10 <- ([#"../12_ghost_code.rs" 33 12 33 19] Len0.len _11);
    goto BB4
  }
  BB4 {
    _9 <- ([#"../12_ghost_code.rs" 33 12 33 24] _10 = (1 : usize));
    _8 <- not _9;
    switch (_8)
      | False -> goto BB6
      | True -> goto BB5
      end
  }
  BB5 {
    absurd
  }
  BB6 {
    _7 <- ();
    _0 <- ();
    goto BB7
  }
  BB7 {
    assume { Resolve0.resolve x_1 };
    return _0
  }
  
end
module C12GhostCode_MyStruct_Type
  use mach.int.Int
  use mach.int.UInt32
  type c12ghostcode_mystruct_type  =
    | C12GhostCode_MyStruct_Type uint32 (ghost uint32)
    
  let ghost function c12ghostcode_mystruct_type_MyStruct_g (self : c12ghostcode_mystruct_type) : uint32
   = [@vc:do_not_keep_trace] [@vc:sp]
    match (self) with
      | C12GhostCode_MyStruct_Type _ a -> a
      end
  let function c12ghostcode_mystruct_type_MyStruct_f (self : c12ghostcode_mystruct_type) : uint32
   = [@vc:do_not_keep_trace] [@vc:sp]
    match (self) with
      | C12GhostCode_MyStruct_Type a _ -> a
      end
end
module C12GhostCode_TakesStruct_Interface
  use mach.int.UInt32
  use mach.int.Int
  use mach.int.Int32
  use C12GhostCode_MyStruct_Type
  val takes_struct [@cfg:stackify] (x : C12GhostCode_MyStruct_Type.c12ghostcode_mystruct_type) : ()
    requires {[#"../12_ghost_code.rs" 41 11 41 20] UInt32.to_int (C12GhostCode_MyStruct_Type.c12ghostcode_mystruct_type_MyStruct_g x) = 0}
    
end
module C12GhostCode_TakesStruct
  use mach.int.UInt32
  use mach.int.Int
  use mach.int.Int32
  use C12GhostCode_MyStruct_Type
  let rec cfg takes_struct [@cfg:stackify] [#"../12_ghost_code.rs" 42 0 42 36] (x : C12GhostCode_MyStruct_Type.c12ghostcode_mystruct_type) : ()
    requires {[#"../12_ghost_code.rs" 41 11 41 20] UInt32.to_int (C12GhostCode_MyStruct_Type.c12ghostcode_mystruct_type_MyStruct_g x) = 0}
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var x_1 : C12GhostCode_MyStruct_Type.c12ghostcode_mystruct_type;
  ghost var _3 : uint32;
  var _6 : ();
  {
    x_1 <- x;
    goto BB0
  }
  BB0 {
    _6 <- ();
    _3 <- ghost ([#"../12_ghost_code.rs" 43 10 43 24] C12GhostCode_MyStruct_Type.c12ghostcode_mystruct_type_MyStruct_f x_1);
    goto BB1
  }
  BB1 {
    x_1 <- (let C12GhostCode_MyStruct_Type.C12GhostCode_MyStruct_Type a b = x_1 in C12GhostCode_MyStruct_Type.C12GhostCode_MyStruct_Type a _3);
    _3 <- any uint32;
    _0 <- ();
    return _0
  }
  
end
