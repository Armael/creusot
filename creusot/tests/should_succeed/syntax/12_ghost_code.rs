extern crate creusot_contracts;
use creusot_contracts::*;

// fn ghost_arg(g: Ghost<u32>) {
//     let x: Ghost<_> = ghost! { g.inner() };
// }

// fn ghost_vec() {
//   let x : Vec<u32> = Vec::new();
//   let mut s : Ghost<_> = ghost! { x };
// }

#[logic]
fn omg() {}

fn ghost_copy() {
    let a = 0;
    let mut s = ghost! { Seq::EMPTY.push(0) };
    s = ghost! { { s.inner().push(a) } };
}

// fn ghost_vec() {
//     let mut v = Vec::new();

//     ghost! { v.push(0) };
// }

#[logic]
fn logi_drop<T>(_: T) {}

fn ghost_check() {
    let mut x = Vec::new();

    // We ghost capture the value and then drop it without affecting program
    ghost! { {logi_drop(x); } };

    x.push(0);

    assert!(x.len() == 1);
}
