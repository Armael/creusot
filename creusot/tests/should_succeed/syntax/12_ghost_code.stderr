warning: unnecessary braces around function argument
  --> 12_ghost_code.rs:19:19
   |
19 |     _s = ghost! { { _s.push(a) } };
   |                   ^^          ^^
   |
   = note: `#[warn(unused_braces)]` on by default
help: remove these braces
   |
19 -     _s = ghost! { { _s.push(a) } };
19 +     _s = ghost! { _s.push(a) };
   |

warning: 1 warning emitted

